keps/sig-node/5328-node-capabilities/README.md

```diff
@@ -0,0 +1,821 @@
+*   If the capability is missing or has a non-matching value, the request is marked as invalid
+
+### Example Walkthrough
+
+**New Features and Version Skew Management**
+
+1. When developing a new Kubelet feature Eg:[ In-Place Pod Resize for Guaranteed QoS pods](https://github.com/kubernetes/enhancements/issues/5294), the skew management strategy involves
+    *   Modifying the Kubelet to advertise the feature as a Node Capability if the feature gate `InPlacePodVerticalScalingExclusiveCPUs` is active. Kubelet would add <code>[kubernetes.io/feature/inPlacePodResizeExclusiveCpus](http://kubernetes.io/feature/inPlacePodResizeExclusiveCpus): true </code>in<code> status.capabilities</code>.
+    *   Updating the <code>NodeCapabilityValidator</code> to check if the node specified in the pod's <code>spec.nodeName</code> possesses this capability before the request is admitted.
+2. An Admin enables the <code>InPlacePodVerticalScalingExclusiveCPUs</code> alpha feature gate on a NodePool. 
+3. The Kubelet on the Node detects this and advertises its support by setting <code>[kubernetes.io/feature/inPlacePodResizeExclusiveCpus](http://kubernetes.io/feature/inPlacePodResizeExclusiveCpus): true </code>in<code> status.capabilities</code>.<code> </code>
+4. An Application Owner has a Guaranteed QoS Pod running on the Node and requests an in place resize. 
+5. When the API server processes the resize request, its validation logic for pod updates is invoked in the admission controller. The request is admitted only if the Node supports the feature. \
+
+
+**Workload Requirements during Scheduling**
+
+1. The feature developer will enhance the Kubelet to conditionally advertise capabilities for `PodLevelResources`. Kubelet will check that both the feature gate is enabled and the underlying OS is Linux before adding the `kubernetes.io/feature/PodLevelResources: "true"` capability to the node's status.
+2. An Application Owner deploys a Pod that requires pod-level resource management by defining resources directly in the pod spec (e.g., `spec.resources.requests`).
+3. The `NodeCapabilityFilter` in kube-scheduler inspects the `PodSpec`. Upon detecting the `spec.resources` field, it infers that the pod requires the `PodLevelResources` capability. The filter automatically rejects all Windows nodes that do not advertise this capability, ensuring the pod is only placed on a compatible Linux nodes.
+
+### Changes on Existing Nodes
+
+Node capabilities should generally represent mostly stable foundational configurations of the Node, not frequently changing real-time statuses (Example: Driver health). Node capabilities are primarily assessed by the kube-scheduler during pod scheduling. Post-scheduling, the Node Capability changes (e.g., kubelet restart with configuration or feature gate changes) is intended to not have any impact on the running pods on the Node and should only be relevant for new workload. 
```

Comment:
> kubelet restart with configuration or feature gate changes

I think the recommendation would be to drain the node before changing the configuration. I think it's safe to assume that doing so without draining the node could have unintended consequences.